core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/service/StreamService.java

```diff
@@ -0,0 +1,89 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.expediagroup.streamplatform.streamregistry.core.service;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Optional;
+
+import lombok.RequiredArgsConstructor;
+
+import org.springframework.stereotype.Component;
+
+import com.expediagroup.streamplatform.streamregistry.core.handler.HandlerWrapper;
+import com.expediagroup.streamplatform.streamregistry.core.predicate.DomainConfiguredEntityPredicateFactory;
+import com.expediagroup.streamplatform.streamregistry.core.predicate.NameDomainPatternMatchPredicateFactory;
+import com.expediagroup.streamplatform.streamregistry.core.predicate.VersionPredicateFactory;
+import com.expediagroup.streamplatform.streamregistry.core.validator.ConfiguredEntityValidator;
+import com.expediagroup.streamplatform.streamregistry.core.validator.DomainConfiguredEntityValidator;
+import com.expediagroup.streamplatform.streamregistry.core.validator.EntityValidator;
+import com.expediagroup.streamplatform.streamregistry.model.Schema;
+import com.expediagroup.streamplatform.streamregistry.model.Stream;
+import com.expediagroup.streamplatform.streamregistry.repository.Repository;
+import com.expediagroup.streamplatform.streamregistry.service.Service;
+
+@Component
+@RequiredArgsConstructor
+public class StreamService implements Service<Stream, Stream.Key> {
+  private final EntityValidator entityValidator;
+  private final ConfiguredEntityValidator configuredEntityValidator;
+  private final DomainConfiguredEntityValidator domainConfiguredEntityValidator;
+  private final HandlerWrapper<Stream> streamHandler;
+  private final Repository<Stream, Stream.Key> streamRepository;
+  private final Repository<Schema, Schema.Key> schemaRepository;
+  private final DomainConfiguredEntityPredicateFactory domainConfiguredEntityPredicateFactory;
+  private final NameDomainPatternMatchPredicateFactory nameDomainPatternMatchPredicateFactory;
+  private final VersionPredicateFactory versionPredicateFactory;
+
+  @Override
+  public void upsert(Stream stream) {
+    Optional<Stream> existing = streamRepository.get(stream.key());
+    entityValidator.validate(stream, existing);
+    configuredEntityValidator.validate(stream, existing);
+    domainConfiguredEntityValidator.validate(stream, existing);
+
+
+    Optional
+        .ofNullable(stream.getVersion())
+        .filter(v -> v > 0)
+        .orElseThrow(() -> new IllegalArgumentException("Version must be provided and greater than 0."));
+    existing.ifPresent(e -> checkArgument(
+        stream.getVersion() > e.getVersion(),
+        "Version greater than the current version."));
```

Comment:
Double check whether the `Stream.Key` includes version, and hence if the logic here is valid.